Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp

```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright (C) 2004, 2005, 2007, 2008 Nikolas Zimmermann <zimmermann@kde.org>
+ * Copyright (C) 2004, 2005, 2006, 2007, 2008 Rob Buis <buis@kde.org>
+ * Copyright (C) Research In Motion Limited 2009-2010. All rights reserved.
+ * Copyright (C) 2011 Dirk Schulze <krit@webkit.org>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public License
+ * along with this library; see the file COPYING.LIB.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "config.h"
+#include "RenderSVGResourceClipper.h"
+
+#include "ElementIterator.h"
+#include "Frame.h"
+#include "FrameView.h"
+#include "HitTestRequest.h"
+#include "HitTestResult.h"
+#include "IntRect.h"
+#include "Logging.h"
+#include "RenderElementInlines.h"
+#include "RenderLayer.h"
+#include "RenderLayerInlines.h"
+#include "RenderSVGResourceClipperInlines.h"
+#include "RenderSVGText.h"
+#include "RenderStyle.h"
+#include "RenderView.h"
+#include "SVGClipPathElement.h"
+#include "SVGContainerLayout.h"
+#include "SVGElementTypeHelpers.h"
+#include "SVGNames.h"
+#include "SVGPathData.h"
+#include "SVGRenderingContext.h"
+#include "SVGResources.h"
+#include "SVGResourcesCache.h"
+#include "SVGUseElement.h"
+#include <wtf/IsoMallocInlines.h>
+#include <wtf/SetForScope.h>
+#include <wtf/text/TextStream.h>
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(RenderSVGResourceContainer);
+
+RenderSVGResourceContainer::RenderSVGResourceContainer(SVGElement& element, RenderStyle&& style)
+    : RenderSVGHiddenContainer(Type::SVGResourceClipper, element, WTFMove(style))
+    , m_id(element.getIdAttribute())
+{
+}
+
+RenderSVGResourceContainer::~RenderSVGResourceContainer() = default;
+
+void RenderSVGResourceContainer::willBeDestroyed()
+{
+    if (m_registered) {
+        treeScopeForSVGReferences().removeSVGResource(m_id);
+        m_registered = false;
+    }
+
+    RenderSVGHiddenContainer::willBeDestroyed();
+}
+
+void RenderSVGResourceContainer::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle)
+{
+    RenderSVGHiddenContainer::styleDidChange(diff, oldStyle);
+
+    if (!m_registered) {
+        m_registered = true;
+        registerResource();
+    }
+}
+
+void RenderSVGResourceContainer::idChanged()
+{
+    // Invalidate all our current clients.
+    removeAllClientsFromCache();
```

Comment:
I'd say we should get rid of these calls, initially, no?

Reply:
Right, removed.